ray-curator/ray_curator/stages/classifiers/aegis.py

```diff
@@ -0,0 +1,555 @@
+        else:
+            return self._forward(batch)
+
+
+class HFAegisModelStage(HFModel):
+    """
+    See HFModel for more information.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        model_identifier: str,
+        hf_token: str | None = None,
+        pred_column: str = "preds",
+        prob_column: str = "probs",
+        micro_batch_size: int = 256,
+        has_seq_order: bool = True,
+        add_instruction_data_guard: bool = False,
+        autocast: bool = True,
+    ):
+        super().__init__(
+            model_identifier=model_identifier,
+            hf_token=hf_token,
+            pred_column=pred_column,
+            micro_batch_size=micro_batch_size,
+            has_seq_order=has_seq_order,
+            padding_side="left",
+        )
+
+        self.add_instruction_data_guard = add_instruction_data_guard
+        self.prob_column = prob_column
+        self.autocast = autocast
+
+    def setup(self, _: WorkerMetadata | None = None) -> None:
+        self.model = AegisModel(
+            pretrained_model_name_or_path="meta-llama/LlamaGuard-7b",
+            peft_model_name_or_path=self.model_identifier,
+            dtype=torch.bfloat16,
+            token=None,
+            add_instruction_data_guard=self.add_instruction_data_guard,
+            autocast=self.autocast,
+        )
+        if self.add_instruction_data_guard:
+            self.model.instruction_data_guard_net = self.model.instruction_data_guard_net.from_pretrained(
+                "nvidia/instruction-data-guard"
+            )
+            self.model.instruction_data_guard_net = self.model.instruction_data_guard_net.cuda().eval()
+
+        self.model = self.model.cuda().eval()
+
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            pretrained_model_name_or_path="meta-llama/LlamaGuard-7b",
+            padding_side="left",
+        )
+        self.tokenizer.pad_token = self.tokenizer.unk_token
+
+    def process_model_output(self, outputs: torch.Tensor) -> dict[str, np.ndarray]:
+        preds = outputs.cpu().numpy()
+        return {
+            "preds": preds,
+        }
+
+    def collect_outputs(self, processed_outputs: list[dict[str, np.ndarray]]) -> dict[str, np.ndarray]:
+        return {
+            "preds": np.concatenate([out["preds"] for out in processed_outputs], axis=0),
+        }
+
+    def create_output_dataframe(self, df_cpu: pd.DataFrame, collected_output: dict[str, np.ndarray]) -> pd.DataFrame:
+        df_cpu = df_cpu.drop(columns=["input_ids", "attention_mask"])
+
+        if self.add_instruction_data_guard:
+            df_cpu[self.prob_column] = collected_output["preds"].tolist()
+            df_cpu[self.pred_column] = (collected_output["preds"] >= 0.5).tolist()  # noqa: PLR2004
+        else:
+            df_cpu[self.pred_column] = collected_output["preds"].tolist()
+
+        return df_cpu
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch:
```

Comment:
The `process` seems to be generic enough, can/should this be abstracted? Also can you create a dumb README.md that shows how to add a new classifier? I always (even in dask) world get confused which classes we had to extend

Reply:
I think I had originally edited it to be different for Aegis, but then reverted back to the underlying `process` function. I can verify that it is okay to fall back on the parent implementation.